### PR TITLE
win/osd: minor performance tuning on malloc

### PR DESCRIPTION
**Pull request by hoopoepg**
- for the most often case (IOV vector where only one element)
  additional temp buffer is not required and may be used in-place
  data buffer
- fix: do not allocate temporary buffer for single-element IOV
- changes are windows specific

Signed-off-by: Oblomov, Sergey <sergey.oblomov@intel.com>

### CODE REVIEW

File: src/windows/osd.c [hunk 1]
```diff
@@ -286,25 +286,31 @@ ssize_t recvmsg(int sd, struct msghdr *msg, int flags)
...
-		len += msg->msg_iov[i].iov_len;
-
-	buffer = (char*)malloc(len);
-	if(!buffer)
-		goto fn_nomem;
+	if (msg->msg_iovlen > 1) {
+		for (i = 0, len = 0; i < msg->msg_iovlen; i++)
+			len += msg->msg_iov[i].iov_len;
+
+		buffer = (char*)malloc(len);
```
Comment (shefty) [inline src/windows/osd.c]: why not use alloca?

**Comment by shefty** [comment]
merged- but see comment on switching from malloc to alloca

File: src/windows/osd.c (hunk 1 above)
Comment (hoopoepg) [inline src/windows/osd.c]: alloca(10GB) may overflow stask :)

**Comment by shefty** [comment]
> +	if (msg->msg_iovlen > 1) {
> +		for (i = 0, len = 0; i < msg->msg_iovlen; i++)
> +			len += msg->msg_iov[i].iov_len;
> +
> +		buffer = (char*)malloc(len);
> 
> alloca(10GB) may overflow stask :)

What size iov is being reported as supported?!

**Comment by hoopoepg** [comment]
max len depends from MTU only, but technically MTU may have any value (1..MAX_INT). for regular IP alloca may be used, but I'm not sure about loopback interfaces (where shared memory segment may be used as transport and MTU there may be really big)

**Comment by shefty** [comment]
What is the iov_limit reported through the fi_info attributes?

**Comment by hoopoepg** [comment]
are you what provider about? here is fix for common API implementation (port from Linux), not provider specific

**Comment by shefty** [comment]
Sorry, I thought this was related to the ND provider.  It's just pure socket code.

**Pull request merged**